coderd/database/dbauthz/dbauthz.go

```diff
@@ -117,29 +117,94 @@ func ActorFromContext(ctx context.Context) (rbac.Subject, bool) {
 	return a, ok
 }
 
-// AsSystem returns a context with a system actor. This is used for internal
-// system operations that are not tied to any particular actor.
-// When you use this function, be sure to add a //nolint comment
-// explaining why it is necessary.
-//
-// We trust you have received the usual lecture from the local System
-// Administrator. It usually boils down to these three things:
-// #1) Respect the privacy of others.
-// #2) Think before you type.
-// #3) With great power comes great responsibility.
-func AsSystem(ctx context.Context) context.Context {
+// AsProvisionerd returns a context with an actor that has permissions required
+// for provisionerd to function.
+func AsProvisionerd(ctx context.Context) context.Context {
+	return context.WithValue(ctx, authContextKey{}, rbac.Subject{
+		ID: uuid.Nil.String(),
+		Roles: rbac.Roles([]rbac.Role{
+			{
+				Name:        "provisionerd",
+				DisplayName: "Provisioner Daemon",
+				Site: rbac.Permissions(map[string][]rbac.Action{
+					rbac.ResourceFile.Type:      {rbac.ActionRead},
+					rbac.ResourceTemplate.Type:  {rbac.ActionRead, rbac.ActionUpdate},
+					rbac.ResourceUser.Type:      {rbac.ActionRead},
+					rbac.ResourceWorkspace.Type: {rbac.ActionRead, rbac.ActionUpdate, rbac.ActionDelete},
+				}),
+				Org:  map[string][]rbac.Permission{},
+				User: []rbac.Permission{},
+			},
+		}),
+		Scope: rbac.ScopeAll,
+	},
+	)
+}
+
+// AsAutostart returns a context with an actor that has permissions required
+// for autostart to function.
+func AsAutostart(ctx context.Context) context.Context {
+	return context.WithValue(ctx, authContextKey{}, rbac.Subject{
+		ID: uuid.Nil.String(),
+		Roles: rbac.Roles([]rbac.Role{
+			{
+				Name:        "autostart",
+				DisplayName: "Autostart Daemon",
+				Site: rbac.Permissions(map[string][]rbac.Action{
+					rbac.ResourceTemplate.Type:  {rbac.ActionRead, rbac.ActionUpdate},
+					rbac.ResourceWorkspace.Type: {rbac.ActionRead, rbac.ActionUpdate},
+				}),
+				Org:  map[string][]rbac.Permission{},
+				User: []rbac.Permission{},
+			},
+		}),
+		Scope: rbac.ScopeAll,
+	},
+	)
+}
+
+// AsMetricsCache returns a context with an actor that has permissions required
+// for the metrics cache to function.
+func AsMetricsCache(ctx context.Context) context.Context {
+	return context.WithValue(ctx, authContextKey{}, rbac.Subject{
+		ID: uuid.Nil.String(),
+		Roles: rbac.Roles([]rbac.Role{
+			{
+				Name:        "metrics-cache",
+				DisplayName: "Metrics Cache",
+				Site: rbac.Permissions(map[string][]rbac.Action{
+					rbac.ResourceTemplate.Type: {rbac.ActionRead},
+				}),
+				Org:  map[string][]rbac.Permission{},
+				User: []rbac.Permission{},
+			},
+		}),
+		Scope: rbac.ScopeAll,
+	},
+	)
+}
+
+// AsSystemRestricted returns a context with an actor that has permissions
+// required for various system operations e.g. login, logout.
+func AsSystemRestricted(ctx context.Context) context.Context {
```

Comment:
Do we plan on keeping this around? Or is it a catch all for the remaining stuff for now?

Reply:
This is a catch-all for the remaining stuff. It's mostly used for HTTP middleware.
I've pared down all the perms except read, which can still be pared down to the bare minimum if need be.

If we need to break it down further in future, we can do so. I think this is fine for now though.